### ISSUE TITLE
Why can the result of `hash` be negative?

### ISSUE THREAD

**Issue by ad-si** [haskell-unordered-containers/hashable#162]
I guess it's an buffer overflow, right?
But why is this allowed to happen?

**Comment by phadej** [comment]
`unordered-containers` uses `Word` as a hash:

https://github.com/tibbe/unordered-containers/blob/efa43a2ab09dc6eb72893d12676a8e188cb4ca63/Data/HashMap/Base.hs#L165-L166

https://github.com/tibbe/unordered-containers/blob/efa43a2ab09dc6eb72893d12676a8e188cb4ca63/Data/HashMap/Base.hs#L233

Report says

> For coercing between any two integer types, use `fromIntegral`. Coercing word types to and from integer types preserves representation, not sign.

So `Word` and `Int` can be interchanged. Also changing result of `hash` to be `Word` would cause more harm that benefit.

TL;DR negative values are completely valid, `hashWithSalt` can use the whole domain of `Int`.

**Comment by ad-si** [comment]
Well, if it's intended behavior, can we at least add a few lines to the documentation informing about this (and also how hash generation actually works in this package)

**Comment by phadej** [comment]
The *description* of the packages says

> This package defines a class, Hashable, for types that can be converted to a hash value. This class exists for the benefit of hashing-based data structures. ...

i.e. other applications (e.g. *hyperloglog*, or any kind of crypto) is out of scope. I recall there were closed issues:
- https://github.com/tibbe/hashable/issues/30
- https://github.com/tibbe/hashable/issues/71
maybe others (I cannot find the one mentioning that there **are not** guarantees that hash values are stable across platforms or library versions), see e.g. https://github.com/jberryman/hashabler (which is used only by `unagi-bloomfilter`)

If Johan's comments from those can be condensed into single paragraph to be added to `description`, that would be great.

**Comment by phadej** [comment]
Note: bloomfilter and hyperloglog are kind of hashing-based data structures, but I'd call them *probabalistic* first. `unordered-containers` are on the other hand *containers*.

**Comment by phadej** [comment]
added a note in https://github.com/haskell-unordered-containers/hashable/pull/205
